core/src/main/java/com/linecorp/armeria/common/stream/StreamMessage.java

```diff
@@ -1073,4 +1074,37 @@ default InputStream toInputStream(Function<? super T, ? extends HttpData> httpDa
         requireNonNull(executor, "executor");
         return new StreamMessageInputStream<>(this, httpDataConverter, executor);
     }
+
+    /**
+     * Dynamically emits the last value depending on whether this {@link StreamMessage} completes successfully
+     * or exceptionally.
+     *
+     * <p>For example:<pre>{@code
+     * StreamMessage<Integer> source = StreamMessage.of(1, 2, 3, 4, 5, 6, 7, 8, 9, 10);
+     * StreamMessage<Integer> aborted = source
+     *     .peek(i -> {
+     *         if (i > 5) {
+     *             source.abort();
+     *         }
+     *     });
+     * StreamMessage<Integer> endWith = aborted
+     *     .endWith(th -> {
+     *          if (th instanceof AbortedStreamException) {
+     *              return 100;
+     *          }
+     *          return -1;
+     *     });
+     * List<Integer> collected = endWith.collect().join();
+     *
+     * assert collected.equals(List.of(1, 2, 3, 4, 5, 100));
+     * }</pre>
+     *
+     * <p>Note that if {@code null} is returned by the {@link Function}, the {@link StreamMessage} will complete
+     * successfully without emitting an additional value when this stream is complete successfully,
+     * or complete exceptionally when this stream complete exceptionally.
```

Comment:
```java
    @Test
    void finalizer_return_null_upstream_thrown() {
        // given
        final StreamMessage<Integer> streamMessage = StreamMessage.of(2, 3, 4, 5);
        final StreamMessage<Integer> aborted = streamMessage
                .peek(val -> {
                    if (val == 5) {
                        throw new RuntimeException();
                    }
                });
        final Function<Throwable, Integer> finalizer = unused -> null;
        final StreamMessage<Integer> publisher = new SurroundingPublisher<>(1, aborted, finalizer);

        // when & then
        StepVerifier.create(publisher)
                    .expectNext(1, 2, 3, 4)
                    .expectError(RuntimeException.class) // 👈 👈 upstreamException is propagated when finalizer return null
                    .verify();
    }
```
@ikhoon , If `finalizer` returns `null` and upstream thrown, `finalize()` propagate exception to downstream with `close(upstreamException)`.

So I updated the comment in more detail~!